Deadlock in TestTriggersConstraints

An unusual failure:
```
Test mvStore traceLevelFile defrag  (23555 KB used)

06:42:14 00:00.326 org.h2.test.db.TestTriggersConstraints FAIL java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
ERROR: FAIL (mvStore traceLevelFile defrag ) java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199] java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199] ------------------------------
java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
	at org.h2.util.Task.get(Task.java:83)
	at org.h2.test.db.TestTriggersConstraints.testTriggerDeadlock(TestTriggersConstraints.java:98)
	at org.h2.test.db.TestTriggersConstraints.test(TestTriggersConstraints.java:45)
	at org.h2.test.TestBase.runTest(TestBase.java:139)
	at org.h2.test.TestAll.addTest(TestAll.java:1030)
	at org.h2.test.TestAll.test(TestAll.java:800)
	at org.h2.test.TestAll.runTests(TestAll.java:682)
	at org.h2.test.TestAll.testAll(TestAll.java:564)
	at org.h2.test.TestAll.run(TestAll.java:514)
	at org.h2.test.TestAll.main(TestAll.java:455)
Caused by: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:461)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:427)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:722)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:229)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:221)
	at org.h2.mvstore.db.MVTable.lockRow(MVTable.java:582)
	at org.h2.command.dml.Update.update(Update.java:125)
	at org.h2.command.CommandContainer.update(CommandContainer.java:133)
	at org.h2.command.Command.executeUpdate(Command.java:267)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:233)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:205)
	at org.h2.test.db.TestTriggersConstraints$1.call(TestTriggersConstraints.java:91)
	at org.h2.util.Task.run(Task.java:44)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Transaction 1 has been chosen as a deadlock victim. Details:
Transaction 2 attempts to update map <table.4> entry with key <1> modified by transaction 1(24) OPEN 3
Transaction 1 attempts to update map <table.5> entry with key <1> modified by transaction 2(25) OPEN 1
 [1.4.199/105]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:894)
	at org.h2.mvstore.tx.Transaction.waitFor(Transaction.java:544)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:319)
	at org.h2.mvstore.tx.TransactionMap.lock(TransactionMap.java:268)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:227)
	... 10 more
```

Once again (currrent `master` branch on Travis):
```Test mvStore net memory

11:08:10 00:00.315 org.h2.test.db.TestTriggersConstraints FAIL java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:463)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:722)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:229)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:221)
	at org.h2.mvstore.db.MVTable.lockRow(MVTable.java:582)
	at org.h2.command.dml.Update.update(Update.java:142)
	at org.h2.command.CommandContainer.update(CommandContainer.java:195)
	at org.h2.command.Command.executeUpdate(Command.java:269)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:406)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:183)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Transaction 1 has been chosen as a deadlock victim. Details:
Transaction 2 attempts to update map <table.3> entry with key <1> modified by transaction 1(20) OPEN 3
Transaction 1 attempts to update map <table.4> entry with key <1> modified by transaction 2(21) OPEN 1
 [1.4.199/105]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:923)
	at org.h2.mvstore.tx.Transaction.waitFor(Transaction.java:544)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:319)
	at org.h2.mvstore.tx.TransactionMap.lock(TransactionMap.java:268)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:227)
	... 8 more
ERROR: FAIL (mvStore net memory ) java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:463)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:722)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:229)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:221)
	at org.h2.mvstore.db.MVTable.lockRow(MVTable.java:582)
	at org.h2.command.dml.Update.update(Update.java:142)
	at org.h2.command.CommandContainer.update(CommandContainer.java:195)
	at org.h2.command.Command.executeUpdate(Command.java:269)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:406)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:183)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Transaction 1 has been chosen as a deadlock victim. Details:
Transaction 2 attempts to update map <table.3> entry with key <1> modified by transaction 1(20) OPEN 3
Transaction 1 attempts to update map <table.4> entry with key <1> modified by transaction 2(21) OPEN 1
 [1.4.199/105]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:923)
	at org.h2.mvstore.tx.Transaction.waitFor(Transaction.java:544)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:319)
	at org.h2.mvstore.tx.TransactionMap.lock(TransactionMap.java:268)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:227)
	... 8 more
 java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:463)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:722)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:229)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:221)
	at org.h2.mvstore.db.MVTable.lockRow(MVTable.java:582)
	at org.h2.command.dml.Update.update(Update.java:142)
	at org.h2.command.CommandContainer.update(CommandContainer.java:195)
	at org.h2.command.Command.executeUpdate(Command.java:269)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:406)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:183)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Transaction 1 has been chosen as a deadlock victim. Details:
Transaction 2 attempts to update map <table.3> entry with key <1> modified by transaction 1(20) OPEN 3
Transaction 1 attempts to update map <table.4> entry with key <1> modified by transaction 2(21) OPEN 1
 [1.4.199/105]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:923)
	at org.h2.mvstore.tx.Transaction.waitFor(Transaction.java:544)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:319)
	at org.h2.mvstore.tx.TransactionMap.lock(TransactionMap.java:268)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:227)
	... 8 more
 ------------------------------
java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:463)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:722)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:229)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:221)
	at org.h2.mvstore.db.MVTable.lockRow(MVTable.java:582)
	at org.h2.command.dml.Update.update(Update.java:142)
	at org.h2.command.CommandContainer.update(CommandContainer.java:195)
	at org.h2.command.Command.executeUpdate(Command.java:269)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:406)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:183)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Transaction 1 has been chosen as a deadlock victim. Details:
Transaction 2 attempts to update map <table.3> entry with key <1> modified by transaction 1(20) OPEN 3
Transaction 1 attempts to update map <table.4> entry with key <1> modified by transaction 2(21) OPEN 1
 [1.4.199/105]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:923)
	at org.h2.mvstore.tx.Transaction.waitFor(Transaction.java:544)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:319)
	at org.h2.mvstore.tx.TransactionMap.lock(TransactionMap.java:268)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:227)
	... 8 more
	at org.h2.util.Task.get(Task.java:83)
	at org.h2.test.db.TestTriggersConstraints.testTriggerDeadlock(TestTriggersConstraints.java:98)
	at org.h2.test.db.TestTriggersConstraints.test(TestTriggersConstraints.java:45)
	at org.h2.test.TestBase.runTest(TestBase.java:140)
	at org.h2.test.TestAll.addTest(TestAll.java:1030)
	at org.h2.test.TestAll.test(TestAll.java:800)
	at org.h2.test.TestAll.runTests(TestAll.java:663)
	at org.h2.test.TestAll.testAll(TestAll.java:564)
	at org.h2.test.TestAll.run(TestAll.java:514)
	at org.h2.test.TestAll.main(TestAll.java:455)
Caused by: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:463)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:722)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:229)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:221)
	at org.h2.mvstore.db.MVTable.lockRow(MVTable.java:582)
	at org.h2.command.dml.Update.update(Update.java:142)
	at org.h2.command.CommandContainer.update(CommandContainer.java:195)
	at org.h2.command.Command.executeUpdate(Command.java:269)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:406)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:183)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Transaction 1 has been chosen as a deadlock victim. Details:
Transaction 2 attempts to update map <table.3> entry with key <1> modified by transaction 1(20) OPEN 3
Transaction 1 attempts to update map <table.4> entry with key <1> modified by transaction 2(21) OPEN 1
 [1.4.199/105]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:923)
	at org.h2.mvstore.tx.Transaction.waitFor(Transaction.java:544)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:319)
	at org.h2.mvstore.tx.TransactionMap.lock(TransactionMap.java:268)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:227)
	... 8 more
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:463)
	at org.h2.engine.SessionRemote.done(SessionRemote.java:610)
	at org.h2.command.CommandRemote.executeUpdate(CommandRemote.java:237)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:233)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:205)
	at org.h2.test.db.TestTriggersConstraints$1.call(TestTriggersConstraints.java:91)
	at org.h2.util.Task.run(Task.java:44)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Current master:
```
Test mvStore  (25555 KB used)

05:37:28 00:00.306 org.h2.test.db.TestTriggersConstraints FAIL java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-200]
ERROR: FAIL (mvStore ) java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-200] java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-200] ------------------------------
java.lang.RuntimeException: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-200]
	at org.h2.util.Task.get(Task.java:83)
	at org.h2.test.db.TestTriggersConstraints.testTriggerDeadlock(TestTriggersConstraints.java:98)
	at org.h2.test.db.TestTriggersConstraints.test(TestTriggersConstraints.java:45)
	at org.h2.test.TestBase.runTest(TestBase.java:143)
	at org.h2.test.TestAll.addTest(TestAll.java:1011)
	at org.h2.test.TestAll.test(TestAll.java:781)
	at org.h2.test.TestAll.runTests(TestAll.java:634)
	at org.h2.test.TestAll.testAll(TestAll.java:558)
	at org.h2.test.TestAll.run(TestAll.java:508)
	at org.h2.test.TestAll.main(TestAll.java:450)
Caused by: org.h2.jdbc.JdbcSQLTransactionRollbackException: Deadlock detected. The current transaction was rolled back. Details: "TEST2"; SQL statement:
update test2 set id = 4 [40001-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:463)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:429)
	at org.h2.message.DbException.get(DbException.java:194)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:711)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:231)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:223)
	at org.h2.mvstore.db.MVTable.lockRow(MVTable.java:567)
	at org.h2.command.dml.Update.update(Update.java:144)
	at org.h2.command.CommandContainer.update(CommandContainer.java:199)
	at org.h2.command.Command.executeUpdate(Command.java:255)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:228)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:201)
	at org.h2.test.db.TestTriggersConstraints$1.call(TestTriggersConstraints.java:91)
	at org.h2.util.Task.run(Task.java:44)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: Transaction 1 has been chosen as a deadlock victim. Details:
Transaction 2 attempts to update map <table.3> entry with key <1> modified by transaction 1(18) OPEN 3
Transaction 1 attempts to update map <table.4> entry with key <1> modified by transaction 2(19) OPEN 1
 [1.4.200/105]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:951)
	at org.h2.mvstore.tx.Transaction.waitFor(Transaction.java:702)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:334)
	at org.h2.mvstore.tx.TransactionMap.lock(TransactionMap.java:282)
	at org.h2.mvstore.db.MVPrimaryIndex.lockRow(MVPrimaryIndex.java:229)
	... 10 more
```

--- a/h2/src/test/org/h2/test/db/TestTriggersConstraints.java
+++ b/h2/src/test/org/h2/test/db/TestTriggersConstraints.java
@@ -12,6 +12,7 @@ import java.sql.SQLException;
 import java.sql.Statement;
 import java.util.Arrays;
 import java.util.HashSet;
+import java.util.concurrent.CountDownLatch;
 import org.h2.api.ErrorCode;
 import org.h2.api.Trigger;
 import org.h2.engine.Session;
@@ -71,44 +72,50 @@ public class TestTriggersConstraints extends TestDb implements Trigger {
     }
 
     private void testTriggerDeadlock() throws Exception {
-        final Connection conn, conn2;
-        final Statement stat, stat2;
-        conn = getConnection("trigger");
-        conn2 = getConnection("trigger");
-        stat = conn.createStatement();
-        stat2 = conn2.createStatement();
-        stat.execute("create table test(id int) as select 1");
-        stat.execute("create table test2(id int) as select 1");
-        stat.execute("create trigger test_u before update on test2 " +
-                "for each row call \"" + DeleteTrigger.class.getName() + "\"");
-        conn.setAutoCommit(false);
-        conn2.setAutoCommit(false);
-        stat2.execute("update test set id = 2");
-        Task task = new Task() {
-            @Override
-            public void call() throws Exception {
-                Thread.sleep(300);
-                stat2.execute("update test2 set id = 4");
-            }
-        };
-        task.execute();
-        Thread.sleep(100);
-        try {
-            stat.execute("update test2 set id = 3");
-            task.get();
-        } catch (SQLException e) {
-            int errorCode = e.getErrorCode();
-            assertTrue(String.valueOf(errorCode),
+        final CountDownLatch latch = new CountDownLatch(2);
+        try (Connection conn = getConnection("trigger")) {
+            Statement stat = conn.createStatement();
+            stat.execute("create table test(id int) as select 1");
+            stat.execute("create table test2(id int) as select 1");
+            stat.execute("create trigger test_u before update on test2 " +
+                    "for each row call \"" + DeleteTrigger.class.getName() + "\"");
+            conn.setAutoCommit(false);
+            stat.execute("update test set id = 2");
+            Task task = new Task() {
+                @Override
+                public void call() throws Exception {
+                    try (Connection conn2 = getConnection("trigger")) {
+                        conn2.setAutoCommit(false);
+                        try (Statement stat2 = conn2.createStatement()) {
+                            latch.countDown();
+                            latch.await();
+                            stat2.execute("update test2 set id = 4");
+                        }
+                        conn2.rollback();
+                    } catch (SQLException e) {
+                        int errorCode = e.getErrorCode();
+                        assertTrue(String.valueOf(errorCode),
+                                ErrorCode.LOCK_TIMEOUT_1 == errorCode ||
+                                ErrorCode.DEADLOCK_1 == errorCode);
+                    }
+                }
+            };
+            task.execute();
+            latch.countDown();
+            latch.await();
+            try {
+                stat.execute("update test2 set id = 3");
+            } catch (SQLException e) {
+                int errorCode = e.getErrorCode();
+                assertTrue(String.valueOf(errorCode),
                         ErrorCode.LOCK_TIMEOUT_1 == errorCode ||
-                        ErrorCode.DEADLOCK_1 == errorCode ||
-                        ErrorCode.COMMIT_ROLLBACK_NOT_ALLOWED == errorCode);
+                        ErrorCode.DEADLOCK_1 == errorCode);
+            }
+            task.get();
+            conn.rollback();
+            stat.execute("drop table test");
+            stat.execute("drop table test2");
         }
-        conn2.rollback();
-        conn.rollback();
-        stat.execute("drop table test");
-        stat.execute("drop table test2");
-        conn.close();
-        conn2.close();
     }
 
     private void testDeleteInTrigger() throws SQLException {

